src/main/java/org/elasticsearch/cluster/routing/RoutingTable.java

```diff
@@ -117,61 +117,39 @@ public RoutingTableValidation validate(MetaData metaData) {
 
     /**
-     * All the shards (replicas) for the provided indices.
+     * All the shards (replicas) for all indices in this routing table.
      *
-     * @param indices The indices to return all the shards (replicas), can be <tt>null</tt> or empty array to indicate all indices
-     * @return All the shards matching the specific index
-     * @throws IndexMissingException If an index passed does not exists
+     * @return All the shards
      */
-    public List<ShardRouting> allShards(String... indices) throws IndexMissingException {
+    public List<ShardRouting> allShards() throws IndexMissingException {
         List<ShardRouting> shards = Lists.newArrayList();
-        if (indices == null || indices.length == 0) {
-            indices = indicesRouting.keySet().toArray(new String[indicesRouting.keySet().size()]);
-        }
+        String[] indices = indicesRouting.keySet().toArray(new String[indicesRouting.keySet().size()]);
         for (String index : indices) {
-            IndexRoutingTable indexRoutingTable = index(index);
-            if (indexRoutingTable == null) {
-                throw new IndexMissingException(new Index(index));
-            }
-            for (IndexShardRoutingTable indexShardRoutingTable : indexRoutingTable) {
-                for (ShardRouting shardRouting : indexShardRoutingTable) {
-                    shards.add(shardRouting);
-                }
-            }
+            List<ShardRouting> allShardsIndex = allShards(index);
+            shards.addAll(allShardsIndex);
         }
         return shards;
     }
 
     /**
-     * All the shards (primary + replicas) for the provided indices grouped (each group is a single element, consisting
-     * of the shard). This is handy for components that expect to get group iterators, but still want in some
-     * cases to iterate over all the shards (and not just one shard in replication group).
+     * All the shards (replicas) for the provided index.
      *
-     * @param indices The indices to return all the shards (replicas), can be <tt>null</tt> or empty array to indicate all indices
-     * @return All the shards grouped into a single shard element group each
-     * @throws IndexMissingException If an index passed does not exists
-     * @see IndexRoutingTable#groupByAllIt()
+     * @param index The index to return all the shards (replicas).
+     * @return All the shards matching the specific index
+     * @throws IndexMissingException If the index passed does not exists
      */
-    public GroupShardsIterator allShardsGrouped(String... indices) throws IndexMissingException {
-        // use list here since we need to maintain identity across shards
-        ArrayList<ShardIterator> set = new ArrayList<>();
-        if (indices == null || indices.length == 0) {
-            indices = indicesRouting.keySet().toArray(new String[indicesRouting.keySet().size()]);
+    public List<ShardRouting> allShards(String index) throws IndexMissingException {
```

Comment:
The diff is missleading here I think. I split `allShards(String... indices)` into `allShards()` and `List<ShardRouting> allShards(String index)`. My idea was to get rid of the ambigous (String... indices) signature to have one method where it is clear that the caller really wants _all indices and one where he can get one single index, and then have the first call the second in a loop (basically splitting the existing code from `allShards(String... indices)`.

I removed `GroupShardsIterator allShardsGrouped(String... indices)` completely because I didn't find any caller in the core ES codebase. The diff just mixed the changes I think, so it is a bit hard to read.